File: bench/kernels/bench_sieve.py
# bench_sieve.py - Sieve of Eratosthenes, N=10,000,000
import time, math

def sieve(N):
    flags = bytearray(b'\x01') * (N + 1)
    flags[0] = flags[1] = 0
    pMax = math.isqrt(N)
    for p in range(2, pMax + 1):
        if flags[p]:
            i = p * p
            while i <= N:
                flags[i] = 0
                i += p
    count = 0
    for i in range(2, N + 1):
        if flags[i]:
            count += 1
    return count

N = 10_000_000
t0 = time.monotonic()
count = sieve(N)
elapsed = int((time.monotonic() - t0) * 1000)
print(f"result: {count}")
print(f"elapsed: {elapsed} ms")

File: bench/kernels/bench_sieve.rb
# bench_sieve.rb - Sieve of Eratosthenes, N=10,000,000
# Ruby has no bytearray; a String is its byte buffer, indexed with setbyte/getbyte.
def sieve(n)
  flags = "\x01" * (n + 1)
  flags.setbyte(0, 0)
  flags.setbyte(1, 0)
  p_max = Integer.sqrt(n)
  p = 2
  while p <= p_max
    if flags.getbyte(p) == 1
      i = p * p
      while i <= n
        flags.setbyte(i, 0)
        i += p
      end
    end
    p += 1
  end
  count = 0
  i = 2
  while i <= n
    count += 1 if flags.getbyte(i) == 1
    i += 1
  end
  count
end

n = 10_000_000
t0 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
result = sieve(n)
elapsed = ((Process.clock_gettime(Process::CLOCK_MONOTONIC) - t0) * 1000).to_i
puts "result: #{result}"
puts "elapsed: #{elapsed} ms"
